File: terminal/terminal.go
// Package terminal provides iTerm2 terminal management for orchestrate.
package terminal

import (
	"fmt"
	"os/exec"
	"strings"
	"time"

	"orchestrate/agents"

	"marwan.io/iterm2"
)

// SessionInfo contains information about a terminal session.
// It can represent either an agent worktree or a custom command.
type SessionInfo struct {
	// For agent sessions
	Path   string // Worktree path
	Branch string // Branch name
	Agent  string // Agent name

	// For custom command sessions
	Command        string // Custom command to run (empty = just open terminal)
	Title          string // Custom title
	ColorR         int    // Custom color R (0-255)
	ColorG         int    // Custom color G (0-255)
	ColorB         int    // Custom color B (0-255)
	WorktreePath   string // Path to worktree (commands run here)
	WorktreeBranch string // Branch name (for display in title)

	// Type indicator
	IsCustomCommand bool
}

// WorktreeInfo is an alias for backward compatibility.
type WorktreeInfo = SessionInfo

// Manager handles iTerm2 window and session management.
type Manager struct {
	app          iterm2.App
	MaxPerWindow int
}

// NewManager creates a new terminal manager connected to iTerm2.
func NewManager(appName string) (*Manager, error) {
	app, err := iterm2.NewApp(appName)
	if err != nil {
		return nil, fmt.Errorf("failed to connect to iTerm2: %w", err)
	}
	return &Manager{
		app:          app,
		MaxPerWindow: 6,
	}, nil
}

// Close closes the iTerm2 connection.
func (m *Manager) Close() {
	if m.app != nil {
		_ = m.app.Close()
	}
}

// MaximizeWindow maximizes the frontmost iTerm window.
func MaximizeWindow() {
	time.Sleep(100 * time.Millisecond) // Wait for window to be ready
	script := `tell application "iTerm2"
		tell current window
			-- Get screen dimensions
			tell application "Finder"
				set screenBounds to bounds of window of desktop
			end tell
			set screenWidth to item 3 of screenBounds
			set screenHeight to item 4 of screenBounds
			-- Set window to fill most of the screen (accounting for dock/menubar)
			set bounds to {0, 25, screenWidth, screenHeight - 50}
		end tell
	end tell`
	_ = exec.Command("osascript", "-e", script).Run()
}

// BuildSessionCommand builds the shell command to run in a session.
func BuildSessionCommand(s SessionInfo, prompt string) string {
	if s.IsCustomCommand {
		return BuildCustomCommand(s)
	}
	return BuildAgentCommand(s, prompt)
}

// BuildAgentCommand builds the command for an agent session.
func BuildAgentCommand(s SessionInfo, prompt string) string {
	// Set session title: "agent: branch-name"
	title := fmt.Sprintf("%s: %s", s.Agent, s.Branch)
	titleCmd := fmt.Sprintf("echo -ne '\\033]0;%s\\007'", title)

	// Set tab color based on agent (appears in tab bar and pane border)
	colorCmd := ""
	if color, ok := agents.GetColor(s.Agent); ok {
		colorCmd = fmt.Sprintf(" && echo -ne '\\033]6;1;bg;red;brightness;%d\\007\\033]6;1;bg;green;brightness;%d\\007\\033]6;1;bg;blue;brightness;%d\\007'", color.R, color.G, color.B)
	}

	// Build full command
	escapedPrompt := strings.ReplaceAll(prompt, "'", "'\\''")
	return fmt.Sprintf("%s%s && cd \"%s\" && %s '%s'\n", titleCmd, colorCmd, s.Path, s.Agent, escapedPrompt)
}

// BuildCustomCommand builds the command for a custom command session.
func BuildCustomCommand(s SessionInfo) string {
	// Build title - include branch if running in a worktree
	title := s.Title
	if title == "" {
		title = "terminal"
	}
	if s.WorktreeBranch != "" {
		title = fmt.Sprintf("[%s] %s", s.WorktreeBranch, title)
	}
	titleCmd := fmt.Sprintf("echo -ne '\\033]0;%s\\007'", title)

	// Set tab color if specified
	colorCmd := ""
	if s.ColorR > 0 || s.ColorG > 0 || s.ColorB > 0 {
		colorCmd = fmt.Sprintf(" && echo -ne '\\033]6;1;bg;red;brightness;%d\\007\\033]6;1;bg;green;brightness;%d\\007\\033]6;1;bg;blue;brightness;%d\\007'", s.ColorR, s.ColorG, s.ColorB)
	}

	// Change to worktree directory
	cdCmd := ""
	if s.WorktreePath != "" {
		cdCmd = fmt.Sprintf(" && cd \"%s\"", s.WorktreePath)
	}

	// Print branch info if in a worktree
	branchCmd := ""
	if s.WorktreeBranch != "" {
		branchCmd = fmt.Sprintf(" && echo '📍 Branch: %s'", s.WorktreeBranch)
	}

	// Build full command - handle empty commands (just open terminal)
	cmd := strings.TrimSpace(s.Command)
	if cmd == "" || cmd == "\\n" {
		// Empty command - just set title, color, cd, and show branch
		return fmt.Sprintf("%s%s%s%s\n", titleCmd, colorCmd, cdCmd, branchCmd)
	}

	return fmt.Sprintf("%s%s%s%s && %s\n", titleCmd, colorCmd, cdCmd, branchCmd, cmd)
}

// SendCommand sends a command to an iTerm2 session.
func SendCommand(session iterm2.Session, command string) error {
	return session.SendText(command)
}

// CreateGridLayout creates a grid of sessions for a batch of worktrees.
// Returns the sessions in order matching the worktrees.
func CreateGridLayout(sessions []iterm2.Session, count int) ([]iterm2.Session, error) {
	if len(sessions) == 0 {
		return nil, fmt.Errorf("no sessions provided")
	}

	var allSessions []iterm2.Session

	switch count {
	case 1:
		allSessions = []iterm2.Session{sessions[0]}
	case 2:
		// Two columns
		left := sessions[0]
		right, _ := left.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		allSessions = []iterm2.Session{left, right}
	case 3:
		// Three columns
		left := sessions[0]
		middle, _ := left.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		right, _ := middle.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		allSessions = []iterm2.Session{left, middle, right}
	case 4:
		// 2x2 grid
		topLeft := sessions[0]
		topRight, _ := topLeft.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		bottomLeft, _ := topLeft.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		bottomRight, _ := topRight.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		allSessions = []iterm2.Session{topLeft, topRight, bottomLeft, bottomRight}
	case 5:
		// 3 on top, 2 on bottom
		topLeft := sessions[0]
		topMiddle, _ := topLeft.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		topRight, _ := topMiddle.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		bottomLeft, _ := topLeft.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		bottomRight, _ := topMiddle.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		allSessions = []iterm2.Session{topLeft, topMiddle, topRight, bottomLeft, bottomRight}
	case 6:
		// 3x2 grid
		topLeft := sessions[0]
		topMiddle, _ := topLeft.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		topRight, _ := topMiddle.SplitPane(iterm2.SplitPaneOptions{Vertical: true})
		bottomLeft, _ := topLeft.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		bottomMiddle, _ := topMiddle.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		bottomRight, _ := topRight.SplitPane(iterm2.SplitPaneOptions{Vertical: false})
		allSessions = []iterm2.Session{topLeft, topMiddle, topRight, bottomLeft, bottomMiddle, bottomRight}
	default:
		// For counts > 6, just return what we have
		allSessions = sessions
	}

	return allSessions, nil
}

// LaunchSessions launches sessions in iTerm2 windows.
func (m *Manager) LaunchSessions(sessions []SessionInfo, prompt string) (int, error) {
	if len(sessions) == 0 {
		return 0, fmt.Errorf("no sessions to launch")
	}

	windowCount := 0

	// Process sessions in batches
	for i := 0; i < len(sessions); i += m.MaxPerWindow {
		// Get batch for this window
		end := i + m.MaxPerWindow
		if end > len(sessions) {
			end = len(sessions)
		}
		batch := sessions[i:end]

		// Create window
		win, err := m.app.CreateWindow()
		if err != nil {
			continue
		}
		MaximizeWindow()
		windowCount++

		tabs, err := win.ListTabs()
		if err != nil || len(tabs) == 0 {
			continue
		}

		itermSessions, err := tabs[0].ListSessions()
		if err != nil || len(itermSessions) == 0 {
			continue
		}

		// Build grid layout
		allSessions, err := CreateGridLayout(itermSessions, len(batch))
		if err != nil {
			continue
		}

		// Send commands to each pane
		for j, session := range allSessions {
			if j < len(batch) {
				cmd := BuildSessionCommand(batch[j], prompt)
				_ = SendCommand(session, cmd)
			}
		}
	}

	return windowCount, nil
}

// LaunchWorktrees launches agents in iTerm2 windows for the given worktrees.
// This is kept for backward compatibility.
func (m *Manager) LaunchWorktrees(worktrees []WorktreeInfo, prompt string) (int, error) {
	return m.LaunchSessions(worktrees, prompt)
}

// WindowCount calculates the number of windows needed for n worktrees.
func (m *Manager) WindowCount(n int) int {
	if n <= 0 {
		return 0
	}
	return (n + m.MaxPerWindow - 1) / m.MaxPerWindow
}

File: src/terminal.rs
use crate::agents;
use anyhow::{anyhow, Context, Result};
use std::process::Command;

#[derive(Debug, Clone)]
pub struct SessionInfo {
    // Agent session
    pub path: Option<String>,
    pub branch: Option<String>,
    pub agent: Option<String>,
    pub activity_log: Option<String>,

    // Custom command session
    pub command: Option<String>,
    pub title: Option<String>,
    pub color_r: Option<u8>,
    pub color_g: Option<u8>,
    pub color_b: Option<u8>,
    pub worktree_path: Option<String>,
    pub worktree_branch: Option<String>,

    pub is_custom_command: bool,
}

impl SessionInfo {
    pub fn agent_session(path: &str, branch: &str, agent: &str) -> Self {
        SessionInfo {
            path: Some(path.to_string()),
            branch: Some(branch.to_string()),
            agent: Some(agent.to_string()),
            activity_log: None,
            command: None,
            title: None,
            color_r: None,
            color_g: None,
            color_b: None,
            worktree_path: None,
            worktree_branch: None,
            is_custom_command: false,
        }
    }

    pub fn custom_command(
        cmd: &str,
        title: &str,
        color: Option<(u8, u8, u8)>,
        worktree_path: &str,
        worktree_branch: &str,
    ) -> Self {
        SessionInfo {
            path: None,
            branch: None,
            agent: None,
            activity_log: None,
            command: Some(cmd.to_string()),
            title: Some(title.to_string()),
            color_r: color.map(|c| c.0),
            color_g: color.map(|c| c.1),
            color_b: color.map(|c| c.2),
            worktree_path: Some(worktree_path.to_string()),
            worktree_branch: Some(worktree_branch.to_string()),
            is_custom_command: true,
        }
    }
}

pub struct TerminalManager {
    pub max_per_window: usize,
    pub maximize_on_launch: bool,
}

impl TerminalManager {
    pub fn new(maximize_on_launch: bool) -> Self {
        TerminalManager {
            max_per_window: 6,
            maximize_on_launch,
        }
    }

    pub fn launch_sessions(&self, sessions: &[SessionInfo], prompt: &str) -> Result<usize> {
        if sessions.is_empty() {
            return Err(anyhow!("no sessions to launch"));
        }

        let mut window_count = 0;
        for chunk in sessions.chunks(self.max_per_window) {
            let script = build_window_script(chunk, prompt);
            run_osascript(&script)?;
            if self.maximize_on_launch {
                maximize_window().ok();
            }
            window_count += 1;
        }

        Ok(window_count)
    }
}

fn osascript_escape(input: &str) -> String {
    input
        .replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
}

fn build_session_command(session: &SessionInfo, prompt: &str) -> String {
    if session.is_custom_command {
        build_custom_command(session)
    } else {
        build_agent_command(session, prompt)
    }
}

fn build_agent_command(session: &SessionInfo, prompt: &str) -> String {
    let title = format!(
        "{}: {}",
        session.agent.clone().unwrap_or_default(),
        session.branch.clone().unwrap_or_default()
    );
    let mut cmd_parts = vec![format!("echo -ne '\\033]0;{}\\007'", title)];

    if let Some(agent) = &session.agent {
        if let Some(color) = agents::get_color(agent) {
            cmd_parts.push(format!(
                "echo -ne '\\033]6;1;bg;red;brightness;{}\\007\\033]6;1;bg;green;brightness;{}\\007\\033]6;1;bg;blue;brightness;{}\\007'",
                color.r, color.g, color.b
            ));
        }
    }

    if let Some(path) = &session.path {
        cmd_parts.push(format!("cd \"{}\"", path));
    }

    let mut escaped_prompt = prompt.replace('\'', "'\\''");
    if escaped_prompt.is_empty() {
        escaped_prompt = prompt.to_string();
    }

    if let Some(agent) = &session.agent {
        if let Some(log) = &session.activity_log {
            cmd_parts.push(format!(
                "LOG=\"{}\"; mkdir -p \"$(dirname \\\"$LOG\\\")\"; touch \"$LOG\"; ( {{ {} '{}'; }} 2>&1 | tee -a \"$LOG\" )",
                log, agent, escaped_prompt
            ));
        } else {
            cmd_parts.push(format!("{} '{}'", agent, escaped_prompt));
        }
    }

    cmd_parts.join(" && ")
}

fn build_custom_command(session: &SessionInfo) -> String {
    let mut title = session
        .title
        .clone()
        .filter(|t| !t.is_empty())
        .unwrap_or_else(|| "terminal".to_string());

    if let Some(branch) = &session.worktree_branch {
        if !branch.is_empty() {
            title = format!("[{}] {}", branch, title);
        }
    }

    let mut cmd_parts = vec![format!("echo -ne '\\033]0;{}\\007'", title)];

    if let (Some(r), Some(g), Some(b)) = (session.color_r, session.color_g, session.color_b) {
        cmd_parts.push(format!(
            "echo -ne '\\033]6;1;bg;red;brightness;{}\\007\\033]6;1;bg;green;brightness;{}\\007\\033]6;1;bg;blue;brightness;{}\\007'",
            r, g, b
        ));
    }

    if let Some(path) = &session.worktree_path {
        cmd_parts.push(format!("cd \"{}\"", path));
    }

    if let Some(branch) = &session.worktree_branch {
        if !branch.is_empty() {
            cmd_parts.push(format!("echo 'Branch: {}'", branch));
        }
    }

    if let Some(cmd) = &session.command {
        let trimmed = cmd.trim();
        if !trimmed.is_empty() && trimmed != "\\n" {
            cmd_parts.push(trimmed.to_string());
        }
    }

    cmd_parts.join(" && ")
}

fn build_window_script(chunk: &[SessionInfo], prompt: &str) -> String {
    let mut lines = Vec::new();
    lines.push("tell application \"iTerm2\"".to_string());
    lines.push("set newWindow to (create window with default profile)".to_string());
    lines.push("set s1 to current session of newWindow".to_string());

    match chunk.len() {
        1 => {}
        2 => {
            lines.push("tell s1 to set s2 to (split vertically with default profile)".to_string());
        }
        3 => {
            lines.push("tell s1 to set s2 to (split vertically with default profile)".to_string());
            lines.push("tell s2 to set s3 to (split vertically with default profile)".to_string());
        }
        4 => {
            lines.push("tell s1 to set s2 to (split vertically with default profile)".to_string());
            lines.push("tell s1 to set s3 to (split horizontally with default profile)".to_string());
            lines.push("tell s2 to set s4 to (split horizontally with default profile)".to_string());
        }
        5 => {
            lines.push("tell s1 to set s2 to (split vertically with default profile)".to_string());
            lines.push("tell s2 to set s3 to (split vertically with default profile)".to_string());
            lines.push("tell s1 to set s4 to (split horizontally with default profile)".to_string());
            lines
                .push("tell s2 to set s5 to (split horizontally with default profile)".to_string());
        }
        _ => {
            // 6 or more (capped at 6)
            lines.push("tell s1 to set s2 to (split vertically with default profile)".to_string());
            lines.push("tell s2 to set s3 to (split vertically with default profile)".to_string());
            lines.push("tell s1 to set s4 to (split horizontally with default profile)".to_string());
            lines
                .push("tell s2 to set s5 to (split horizontally with default profile)".to_string());
            lines
                .push("tell s3 to set s6 to (split horizontally with default profile)".to_string());
        }
    }

    let pane_order = match chunk.len() {
        1 => vec!["s1"],
        2 => vec!["s1", "s2"],
        3 => vec!["s1", "s2", "s3"],
        4 => vec!["s1", "s2", "s3", "s4"],
        5 => vec!["s1", "s2", "s3", "s4", "s5"],
        _ => vec!["s1", "s2", "s3", "s4", "s5", "s6"],
    };

    for (idx, pane) in pane_order.iter().enumerate() {
        if idx >= chunk.len() {
            break;
        }
        let cmd = osascript_escape(&build_session_command(&chunk[idx], prompt));
        lines.push(format!("tell {} to write text \"{}\"", pane, cmd));
    }

    lines.push("end tell".to_string());
    lines.join("\n")
}

fn run_osascript(script: &str) -> Result<()> {
    let output = Command::new("osascript")
        .arg("-e")
        .arg(script)
        .output()
        .context("failed to run osascript")?;

    if !output.status.success() {
        return Err(anyhow!(
            "osascript failed: {}",
            String::from_utf8_lossy(&output.stderr)
        ));
    }
    Ok(())
}

pub fn focus_worktree_window(worktree_path: &str) -> Result<bool> {
    let script = format!(
        r#"
        tell application "iTerm2"
            set foundWindow to missing value
            repeat with aWindow in windows
                repeat with aTab in tabs of aWindow
                    repeat with aSession in sessions of aTab
                        try
                            set sessionPath to (tty of aSession) as text
                            if sessionPath contains "{}" then
                                set foundWindow to aWindow
                                exit repeat
                            end if
                        end try
                    end repeat
                    if foundWindow is not missing value then exit repeat
                end repeat
                if foundWindow is not missing value then exit repeat
            end repeat
            if foundWindow is not missing value then
                select foundWindow
                tell application "iTerm2" to activate
                return true
            else
                return false
            end if
        end tell
        "#,
        worktree_path
    );

    let output = Command::new("osascript").arg("-e").arg(script).output()?;
    let result = String::from_utf8_lossy(&output.stdout).trim().to_string();
    Ok(result == "true")
}

fn maximize_window() -> Result<()> {
    let script = r#"
        tell application "iTerm2"
            tell current window
                tell application "Finder"
                    set screenBounds to bounds of window of desktop
                end tell
                set bounds to {0, 25, (item 3 of screenBounds), (item 4 of screenBounds) - 50}
            end tell
        end tell
    "#;
    run_osascript(script)
}

pub fn focus_worktree_window_by_branch(branch: &str) -> Result<bool> {
    let script = format!(
        r#"
        tell application "iTerm2"
            set foundWindow to missing value
            repeat with aWindow in windows
                repeat with aTab in tabs of aWindow
                    repeat with aSession in sessions of aTab
                        try
                            set sessionName to (name of aSession) as text
                            if sessionName contains "{}" then
                                set foundWindow to aWindow
                                exit repeat
                            end if
                        end try
                    end repeat
                    if foundWindow is not missing value then exit repeat
                end repeat
                if foundWindow is not missing value then exit repeat
            end repeat
            if foundWindow is not missing value then
                select foundWindow
                tell application "iTerm2" to activate
                return true
            else
                return false
            end if
        end tell
        "#,
        branch
    );

    let output = Command::new("osascript").arg("-e").arg(script).output()?;
    let result = String::from_utf8_lossy(&output.stdout).trim().to_string();
    Ok(result == "true")
}
